### ISSUE TITLE
Can this work on WASM?

### ISSUE THREAD

**Issue by RubenVerg** [remiturk/libffi#15]
Does this library work on the WebAssembly backend for GHC? As far as I can tell it comes with its own version of libffi, but maybe it's incompatible with the code in this repo? Here's the error (using GHC WASM 9.10.1):

```
Internal.hsc:37:32: error: use of undeclared identifier 'FFI_TYPE_STRUCT'
   37 |         static int test_array[(FFI_TYPE_STRUCT) > 0 ? 2 : 1];
      |                                ^
Internal.hsc:38:32: error: no member named 'elements' in 'ffi_type'
   38 |         static int test_array[(offsetof(ffi_type, elements)) > 0 ? 2 : 1];
      |                                ^                  ~~~~~~~~
/home/rubenverg/.ghc-wasm/wasi-sdk/lib/clang/19/include/__stddef_offsetof.h:16:24: note: expanded from macro 'offsetof'
   16 | #define offsetof(t, d) __builtin_offsetof(t, d)
      |                        ^                     ~
```

I presume a version of libffi exists because the error isn't "libffi header not found" but "`FFI_TYPE_STRUCT` not found" and `ffi_type` seems to exist too.

**Comment by RyanGlScott** [comment]
> Does this library work on the WebAssembly backend for GHC?

Hard to say, as I've never tried. Nor do I have a Wasm-enabled GHC compile handy, so it's not easy for me to test locally.

> As far as I can tell it comes with its own version of libffi, but maybe it's incompatible with the code in this repo?

That's quite possible. I feel very out of my depth trying to debug Wasm-related issues, but I'd be happy to review a patch that adds any Wasm-specific tweaks necessary to make the library work.

**Comment by RubenVerg** [comment]
After further investigation, it looks like `dlsym` doesn't work on WASM either, so there's no point to getting the lib to work.

**Comment by RyanGlScott** [comment]
Indeed, I would be surprised if Wasm had a direct counterpart to `dlsym()` or `LoadLibrary()`. That being said, there might be a more Wasm-friendly way to achieve the same thing. Some searching turns up this blog post by the maintainer of the GHC Wasm backend:

* https://www.tweag.io/blog/2022-03-17-libffi-wasm32/

There, he mentions https://gitlab.haskell.org/ghc/libffi-wasm, a code generator that can generate Wasm-friendly `libffi` functionality. Admittedly, I haven't tried this myself, but that seems like a good place to start from.

Once we collectively understand how this works better, we can think about whether it makes sense to integrate it with this library in some way, or if we'd be better served approaching the problem differently from how this library does it.

**Comment by RubenVerg** [comment]
As far as I can tell, that library is the one that ships with GHC WASM and therefore gets used by this library. I'm not sure there's a way to use external libraries though; it looks like WASI doesn't support `dlopen` and `dlsym`; so it probably only works for the GHC-internal use, or maybe a completely different strategy is required.
